votor/src/staked_validators_cache.rs

```diff
@@ -0,0 +1,599 @@
+            .unwrap_or_else(|| {
+                error!(
+                    "StakedValidatorsCache::get: unknown Bank::epoch_staked_nodes for epoch: \
+                     {epoch}"
+                );
+                Arc::<HashMap<Pubkey, u64>>::default()
+            });
+
+        struct Node {
+            pubkey: Pubkey,
+            stake: u64,
+            alpenglow_socket: SocketAddr,
+        }
+
+        let mut nodes: Vec<_> = epoch_staked_nodes
+            .iter()
+            .filter(|(pubkey, stake)| {
+                let positive_stake = **stake > 0;
+                let not_self = pubkey != &&cluster_info.id();
+
+                positive_stake && (self.include_self || not_self)
+            })
+            .filter_map(|(pubkey, stake)| {
+                cluster_info.lookup_contact_info(pubkey, |node| {
+                    node.alpenglow().map(|alpenglow_socket| Node {
+                        pubkey: *pubkey,
+                        stake: *stake,
+                        alpenglow_socket,
+                    })
+                })?
+            })
+            .collect();
+
+        nodes.dedup_by_key(|node| node.alpenglow_socket);
+        nodes.sort_unstable_by(|a, b| a.stake.cmp(&b.stake));
+
+        let mut alpenglow_sockets = Vec::new();
+        let override_map = self
+            .alpenglow_port_override
+            .as_ref()
+            .map(|x| x.get_override_map());
+        for node in nodes {
+            let alpenglow_socket = node.alpenglow_socket;
+            let socket = if let Some(override_map) = &override_map {
+                // If we have an override, use it.
+                override_map
+                    .get(&node.pubkey)
+                    .cloned()
+                    .unwrap_or(alpenglow_socket)
+            } else {
+                alpenglow_socket
+            };
+            alpenglow_sockets.push(socket);
+        }
+        self.cache.push(
+            epoch,
+            StakedValidatorsCacheEntry {
+                alpenglow_sockets,
+                creation_time: update_time,
+            },
+        );
+    }
+
+    pub fn get_staked_validators_by_slot(
+        &mut self,
+        slot: Slot,
+        cluster_info: &ClusterInfo,
+        access_time: Instant,
+    ) -> (&[SocketAddr], bool) {
+        // Check if self.alpenglow_port_override has a different last_modified.
+        // Immediately refresh the cache if it does.
+        if let Some(alpenglow_port_override) = &self.alpenglow_port_override {
+            if alpenglow_port_override.has_new_override(self.alpenglow_port_override_last_modified)
+            {
+                self.alpenglow_port_override_last_modified =
+                    alpenglow_port_override.last_modified();
+                trace!(
+                    "refreshing cache entry for epoch {} due to alpenglow port override \
+                     last_modified change",
+                    self.cur_epoch(slot)
```

Comment:
this grabs a read lock under the hood. Should we grab current epoch once and re-use it?

Reply:
If you are talking about epoch_schedule, I believe we can cache it, since it doesn't change after genesis. Changed, does this look better?